// roots/lib/api/new.js
// Generated by CoffeeScript 1.7.0
(function() {
  var EventEmitter, New, exec, fs, global_config, nodefn, path, sprout, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  path = require('path');

  fs = require('fs');

  EventEmitter = require('events').EventEmitter;

  exec = require('child_process').exec;

  nodefn = require('when/node');

  sprout = require('sprout');

  global_config = require('../global_config');

  _ = require('lodash');

  New = (function(_super) {
    var has_deps, init, install_deps;

    __extends(New, _super);

    function New(roots) {
      this.roots = roots;
      this.base_url = 'https://github.com/roots-dev/base.git';
    }

    New.prototype.exec = function(opts) {
      this.path = opts.path || (function() {
        throw new Error('missing path');
      })();
      this.template = opts.template || global_config().get('default_template');
      this.options = opts.options;
      if (!_.contains(sprout.list(), 'roots-base')) {
        sprout.add({
          name: 'roots-base',
          url: this.base_url
        })["catch"]((function(_this) {
          return function(err) {
            return _this.emit('error', err);
          };
        })(this)).tap((function(_this) {
          return function() {
            return _this.emit('template:base_added');
          };
        })(this)).then((function(_this) {
          return function() {
            return init.call(_this);
          };
        })(this));
      } else {
        init.call(this);
      }
      return this;
    };

    init = function() {
      return sprout.init({
        template: this.template,
        path: this.path,
        options: this.options
      }).tap((function(_this) {
        return function() {
          return _this.emit('template:created');
        };
      })(this)).then((function(_this) {
        return function() {
          if (has_deps.call(_this)) {
            return install_deps.call(_this);
          }
        };
      })(this)).done(((function(_this) {
        return function() {
          return _this.emit('done', _this.path);
        };
      })(this)), ((function(_this) {
        return function(err) {
          return _this.emit('error', err);
        };
      })(this)));
    };

    has_deps = function() {
      return fs.existsSync(path.join(this.path, 'package.json'));
    };

    install_deps = function() {
      this.emit('deps:installing');
      return nodefn.call(exec, "cd " + this.path + " && npm install").tap((function(_this) {
        return function() {
          return _this.emit('deps:finished');
        };
      })(this));
    };

    return New;

  })(EventEmitter);

  module.exports = New;


  /*
  
  What's Going On Here?
  ---------------------
  
  The 'new' class handles the creation of roots templates. It exposes an event emitter, but it's internal methods operate entirely via promises to keep the internal code clean and expose hooks to each part of the process.
  
  The main method, exec, will check to see if the user has any templates installed. If not, it will add the base template, which is the roots default, then continue. If there are already templates, it will continue, assuming the user has used roots before. It will then initialize the template through sprout and return a promise for when it has finished.
  
  The second, has_deps is a simple method that checks to see if the project has a package.json and therefore dependencies that need to be installed. It returns a boolean.
  
  Finally, install_deps will jump into the project and run `npm install`. Throughout the process, events are emitted at times where it might be useful to be able to hook into  the process via the public API.
  
  To see a sample implementation, check out commands/new.coffee to see how you can hook into the events.
   */

}).call(this);
