// roots/lib/api/watch.js
// Generated by CoffeeScript 1.7.0
(function() {
  var Watcher, chokidar, minimatch;

  chokidar = require('chokidar');

  minimatch = require('minimatch');

  Watcher = (function() {
    var ignore;

    function Watcher(roots) {
      this.roots = roots;
    }

    Watcher.prototype.exec = function() {
      return this.roots.compile().once('done', (function(_this) {
        return function() {
          return chokidar.watch(_this.roots.root, {
            ignoreInitial: true,
            ignored: ignore.bind(_this)
          }).on('error', function(err) {
            return _this.roots.emit('error', err);
          }).on('change', _this.roots.compile.bind(_this.roots));
        };
      })(this));
    };

    ignore = function(p) {
      var f;
      f = p.replace(this.roots.root, '').slice(1);
      return this.roots.config.watcher_ignores.map(function(i) {
        return minimatch(f, i, {
          dot: true
        });
      }).filter(function(i) {
        return i;
      }).length;
    };

    return Watcher;

  })();

  module.exports = Watcher;

}).call(this);
