// roots/lib/extensions/compiled.js
// Generated by CoffeeScript 1.7.0
(function() {
  var path, _;

  _ = require('lodash');

  path = require('path');

  module.exports = function() {

    /**
     * @class
     * @classdesc This extension puts files into a "compiled" category if their
     * extensions match to an extension that an installed compiler is looking for
     */
    var Compiled;
    return Compiled = (function() {
      var detect_fn;

      function Compiled(roots) {
        this.roots = roots;
        this.category = 'compiled';
      }

      Compiled.prototype.fs = function() {
        return {
          extract: true,
          ordered: true,
          detect: detect_fn.bind(this)
        };
      };

      detect_fn = function(f) {
        var exts;
        exts = _(this.roots.config.compilers).map(function(i) {
          return i.extensions;
        }).flatten().value();
        return _.contains(exts, path.extname(f.relative).slice(1));
      };

      return Compiled;

    })();
  };

}).call(this);
